docs/src/proposals/blockstore-rocksdb-compaction.md

```diff
@@ -0,0 +1,234 @@
+amplification with no write stalls.
+
+### FIFO Compaction
+FIFO Compaction is one of the three compaction styles offered by RocksDB which
+actually does not run any compaction in background.  Instead, it simply throw
+aways files from old to new when one of the following conditions meets:
+
+- If TTL is set for the FIFO Compaction, RocksDB will throw away files which is
+  older than the configured TTL.
+
+- In case the total DB size is still greater than the configured size, RocksDB
+  will throw away files from the oldest to the newest until the total DB size
+  is under the configured size.
+
+Since FIFO Compaction will throw away files without even taking a look, both
+the TTL and the total DB size must be configured with extra buffer (for
+instance, reserve extra 25% in addition to the target TTL / DB size.)
+
+### Fully Utilize Storage by Properly Rate Limiting
+As mentioned previously, FIFO Compaction deletes the oldest SST file when 1)
+its file creation time is than the specified TTL, or 2) when the total DB size
+exceeds the configured size.  While the TTL condition is what we hope to
+happen constantly, we should avoid trigging the size condition whenever
+possible as it the total amount of data within TTL exceeds the storage size.
+
+To avoid data loss happens due to the size condition, we should set rate
+limiting on writes by `storage size / TTL`.  This will ensure Solana's
+BlockStore handles the write traffic just as fast as how it can digest
+without data loss.
+
+### Amplification Factors
+FIFO Compaction offers constant write and space amplifications as it does
+not run any compactions in background.  However, FIFO Compaction usually have
+large a read amplification, as each SST file is usually a logical level,
+and the read amplification is the same as the number of SST files.
+
+Luckily, it is not the case in Solana's BlockStore workload because write-keys
+are mostly monotonically increasing over time.  Thus, the resulting SST files
+will mostly have disjoint key range.  For most reads, it can be answered by
+hitting mostly one SST file or at most two SST files in case there're some
+small out-of-order writes.
+
+To sum up, if no other manual compaction is issued for quickly picking up
+deletions, FIFO Compaction offers the following amplification factors
+in Solana's BlockStore use case:
+
+- Write Amplification: 1 (all data is written once without compaction.)
+- Read Amplification: < 1.1 (assuming each SST file has 10% overlapping key
+  range with another SST file.)
+- Space Amplification: 1 (same data never be written in more than one SST file,
+  and no additional temporary space required for compaction.)
+
+### Compaction Style Migration
+As FIFO Compaction assumes each SST file represents one logical level, RocksDB
+instances compacted by other compaction styles are valid FIFO compacted
+instances.  Migration from Level Compaction to FIFO Compaction should be simple.
```

Comment:
No.  It will require running manual compaction first (via CompactRange) in order to produce logical levels which level compaction work with.  Otherwise, level compaction will treat every file as L0 file and we will immediately hit write stall,

Reply:
Got it - thanks!